metrics/graphite/graphite.go

```diff
@@ -0,0 +1,295 @@
+
+	// buffer the writer and make sure to flush it
+	w := bufio.NewWriter(conn)
+	defer w.Flush()
+
+	now := time.Now().Unix()
+
+	// emit counter stats
+	for _, c := range e.counters {
+		fmt.Fprintf(w, "%s.%s.count %d %d\n", e.prefix, c.Name(), c.count, now)
+	}
+
+	// emit histogram specific stats
+	for _, h := range e.histograms {
+		hist := h.hist.Merge()
+		fmt.Fprintf(w, "%s.%s.count %d %d\n", e.prefix, h.Name(), hist.TotalCount(), now)
+		fmt.Fprintf(w, "%s.%s.min %d %d\n", e.prefix, h.Name(), hist.Min(), now)
+		fmt.Fprintf(w, "%s.%s.max %d %d\n", e.prefix, h.Name(), hist.Max(), now)
+		fmt.Fprintf(w, "%s.%s.mean %.2f %d\n", e.prefix, h.Name(), hist.Mean(), now)
+		fmt.Fprintf(w, "%s.%s.std-dev %.2f %d\n", e.prefix, h.Name(), hist.StdDev(), now)
+	}
+
+	// emit gauge stats (which can include some histogram quantiles)
+	for _, g := range e.gauges {
+		fmt.Fprintf(w, "%s.%s %.2f %d\n", e.prefix, g.Name(), g.Get(), now)
+	}
+}
+
+type counter struct {
+	key   string
+	count uint64
+}
+
+func (c *counter) Name() string { return c.key }
+
+func (c *counter) With(metrics.Field) metrics.Counter { return c }
+
+func (c *counter) Add(delta uint64) { atomic.AddUint64(&c.count, delta) }
+
+type gauge struct {
+	key   string
+	value uint64 // math.Float64bits
+}
+
+func (g *gauge) Name() string { return g.key }
+
+func (g *gauge) With(metrics.Field) metrics.Gauge { return g }
+
+func (g *gauge) Add(delta float64) {
+	for {
+		old := atomic.LoadUint64(&g.value)
+		new := math.Float64bits(math.Float64frombits(old) + delta)
+		if atomic.CompareAndSwapUint64(&g.value, old, new) {
+			return
+		}
+	}
+}
+
+func (g *gauge) Set(value float64) {
+	atomic.StoreUint64(&g.value, math.Float64bits(value))
+}
+
+func (g *gauge) Get() float64 {
+	return math.Float64frombits(atomic.LoadUint64(&g.value))
+}
+
+type windowedHistogram struct {
+	mu   sync.Mutex
+	hist *hdrhistogram.WindowedHistogram
+
+	name   string
+	gauges map[int]metrics.Gauge
+}
+
+// newWindowedHistogram is taken from http://github.com/codahale/metrics. It returns a
+// windowed HDR histogram which drops data older than five minutes.
+//
+// The histogram exposes metrics for each passed quantile as gauges. Users are expected
+// to provide their own set of Gauges for quantiles to make this Histogram work across multiple
+// metrics providers.
+func newWindowedHistogram(name string, minValue, maxValue int64, sigfigs int, quantiles map[int]metrics.Gauge) *windowedHistogram {
+	h := &windowedHistogram{
+		hist:   hdrhistogram.NewWindowed(5, minValue, maxValue, sigfigs),
+		name:   name,
+		gauges: quantiles,
+	}
+	go h.rotateLoop(1 * time.Minute)
+	return h
+}
+
+func (h *windowedHistogram) Name() string                         { return h.name }
+func (h *windowedHistogram) With(metrics.Field) metrics.Histogram { return h }
+
+func (h *windowedHistogram) Observe(value int64) {
+	h.mu.Lock()
+	err := h.hist.Current.RecordValue(value)
+	h.mu.Unlock()
+
+	if err != nil {
+		panic(err.Error())
```

Comment:
Is this truly panic-worthy? Perhaps an alternative would be to log an error and drop the observation?